compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/Types.java

```diff
@@ -2900,18 +2900,63 @@ private boolean checkFillerValue(BFiniteType type) {
         return defaultFillValuePresent;
     }
 
+    private boolean hasImplicitDefaultValue(BFiniteType finiteType) {
+        for (BLangExpression expression : finiteType.getValueSpace()) {
+            if (isImplicitDefaultValue(expression)) {
+                return true;
+            }
+        }
+        return false;
+    }
+
     private boolean checkFillerValue(BUnionType type) {
         if (type.isNullable()) {
             return true;
         }
-        Iterator<BType> iterator = type.getMemberTypes().iterator();
+
+        Set<BType> memberTypes = new HashSet<>();
+        boolean hasFillerValue = false;
+        boolean defaultValuePresent = false;
+        boolean valueTypePresent = false;
+        for (BType member : type.getMemberTypes()) {
+            if (member.tag == TypeTags.FINITE) {
+                Set<BType> uniqueValues = getUniqueValues((BFiniteType) member);
+                memberTypes.addAll(uniqueValues);
+                if (!defaultValuePresent && hasImplicitDefaultValue((BFiniteType) member)) {
+                    defaultValuePresent = true;
+                }
+                valueTypePresent = true;
+            } else {
+                memberTypes.add(member);
+            }
+            if (!hasFillerValue && hasFillerValue(member)) {
+                hasFillerValue = true;
+            }
+        }
+        if (!hasFillerValue) {
+            return false;
+        }
+
+        Iterator<BType> iterator = memberTypes.iterator();
         BType firstMember = iterator.next();
         while (iterator.hasNext()) {
             if (!isSameType(firstMember, iterator.next())) {
                 return false;
             }
         }
-        return isValueType(firstMember) && hasFillerValue(firstMember);
+
+        if (valueTypePresent) {
+            return defaultValuePresent;
+        }
+        return false;
```

Comment:
Is it correct to return false here?

I don't think the following will work.

```ballerina
import ballerina/lang.'int;

public function main() {
    ('int:Unsigned16|'int:Signed32)[2] x = [];
}
```

Reply:
Fixed by a88df61403faefa21fcc66a2ff65ebb348e4f63a